agentrs/src/ws/client.rs

```diff
@@ -0,0 +1,190 @@
+        let config_manager =
+            conf::ConfigHandleManager::init().context("Failed to init config manager")?;
+
+        let gateway_url = std::env::var("GATEWAY_URL")
+            .unwrap_or_else(|_| "ws://localhost:8009/api/ws".to_string());
+        let mut request = gateway_url.into_client_request().unwrap();
+
+        // Insert a custom header
+        let token = config_manager.conf.token.clone().unwrap();
+        let token: &'static str = Box::leak(token.into_boxed_str());
+        request
+            .headers_mut()
+            .insert("HOOP_KEY", HeaderValue::from_static(token));
+
+        Ok(WebSocket {
+            request,
+            config_manager,
+            reconnect_interval: Duration::from_secs(5),
+        })
+    }
+
+    async fn run_with_reconnect(self) -> anyhow::Result<()> {
+        let mut attempts = 0;
+        loop {
+            match self.clone().run().await {
+                Ok(_) => {
+                    debug!("> WebSocket connection closed gracefully");
+                    return Ok(());
+                }
+                Err(e) if e.to_string().contains("401 Unauthorized") => {
+                    error!("> Unauthorized: Invalid token provided. Please check your HOOP_KEY.");
+                    return Err(e);
+                }
+                Err(e) => {
+                    attempts += 1;
+                    error!("> Connection failed (attempt {}): {}", attempts, e);
+
+                    tokio::time::sleep(self.reconnect_interval).await;
+                    continue;
+                }
+            }
+        }
+    }
+
+    async fn run(self) -> anyhow::Result<()> {
+        let (ws_stream, _) = connect_async(self.request.clone()).await?;
+
+        let (ws_sender, ws_receiver) = ws_stream.split();
+
+        // Clone config manager and sessions for use in the async task
+        let ws_sender = Arc::new(Mutex::new(ws_sender));
+        let sessions: SessionMap = Arc::new(RwLock::new(HashMap::new()));
+        // Store active RDP proxy tasks per session
+        let active_proxies: ProxyMap = Arc::new(RwLock::new(HashMap::new()));
+        let session_channels: ChannelMap = Arc::new(RwLock::new(HashMap::new()));
+
+        let message_processor = MessageProcessor {
+            ws_sender: ws_sender.clone(),
+            sessions: sessions.clone(),
+            active_proxies: active_proxies.clone(),
+            session_channels: session_channels.clone(),
+            config_manager: self.config_manager.clone(),
+        };
+
+        let processor_task =
+            tokio::spawn(async move { message_processor.process_messages(ws_receiver).await });
+
+        // Start heartbeat task in case connection stucked or deadlock
+        let heartbeat_task = self.spawn_heartbeat_task(ws_sender.clone());
+
+        let result = tokio::select! {
+            result = processor_task => {
+                match result {
+                    Ok(Ok(())) => {
```

Comment:
Just an idea, maybe I would make a `match` inside the `Ok` case

Reply:
same i will do this later